content/java/concepts/compiler/compiler.md

```diff
@@ -0,0 +1,41 @@
+---
+Title: 'Compiler'
+Description: 'Java compilers typically take Java source code and produce class files containing platform-neutral Java bytecode that can be executed by the Java Virtual Machine (JVM).'
+Subjects:
+  - 'Code Foundations'
+  - 'Computer Science'
+Tags:
+  - 'Command Line'
+  - 'Developer Tools'
+CatalogContent:
+  - 'learn-java'
+  - 'paths/computer-science'
+---
+
+Java **compilers** are programs that take Java source code and produce class files containing platform-neutral Java bytecode that can be executed by the Java Virtual Machine (JVM).
+
+The JVM does not interpret source code (high-level program). Instead, it interprets what is called Java bytecode (a low-level language somewhere between human-readable source code, and [machine code](https://www.codecademy.com/resources/docs/general/machine-code)) which is specific to a particular computer. Unlike machine code, bytecode is platform-neutral meaning the same bytecode can be interpreted by any JVM running on any sort of computer system. This is what makes compiled Java programs portable.
+
+Most of these compilers do little to no optimization of the code, leaving that task for the JVM to do at run time. The JVM loads the bytecode and either interprets it, or just-in-time (JIT) compiles it to machine code, and then possibly optimizes it.
```

Comment:
```suggestion
Rather than interpret high-level Java source code, the JVM interprets low-level Java bytecode (somewhere between human-readable code and [machine code](https://www.codecademy.com/resources/docs/general/machine-code)) that is specific to a particular computer. Bytecode is platform-neutral and, therefore, can be interpreted by any JVM running on any computer system. This is what makes compiled Java programs portable.

Most of these compilers do little to no optimization of the code, leaving that task for the JVM to do at run time. The JVM loads the bytecode and either interprets it, or just-in-time (JIT) compiles it to machine code, and then possibly optimizes it.
```

I'm a bit confused by the sentence, "Most of these compilers do little to no optimization of the code, leaving that task for the JVM to do at run time." Which compilers are we referring to? And does "possibly" mean there are edge cases where JVM will not optimize the code? If so, what are 1-2 cases of this?